File: frontend/src/state/diagnostics.ts
import type { SerializedDiagnostic } from "../transport/dto";

export type DiagnosticFilter = "all" | SerializedDiagnostic["severity"];

export interface DiagnosticPresentation extends SerializedDiagnostic {
  id: string;
}

export interface DiagnosticSummary {
  all: number;
  error: number;
  warning: number;
  info: number;
}

const severityOrder = {
  error: 0,
  warning: 1,
  info: 2,
} satisfies Record<SerializedDiagnostic["severity"], number>;

function compareText(left: string, right: string) {
  if (left === right) {
    return 0;
  }
  return left < right ? -1 : 1;
}

function detail(
  diagnostic: SerializedDiagnostic,
  key: string,
  fallback = "",
) {
  return diagnostic.details[key] ?? fallback;
}

export function localizedDiagnosticMessage(
  diagnostic: SerializedDiagnostic,
): string {
  switch (diagnostic.code) {
    case "required-metadata":
      return `缺少必填元数据：${detail(
        diagnostic,
        "path",
        diagnostic.target ?? "",
      )}`;
    case "empty-document":
      return "论文正文为空";
    case "invalid-id-prefix":
      return `ID 前缀无效：${diagnostic.target ?? ""}，期望 ${detail(
        diagnostic,
        "expected",
      )}`;
    case "duplicate-id":
      return `重复 ID：${diagnostic.target ?? ""}`;
    case "missing-reference":
      return `引用目标不存在：${diagnostic.target ?? ""}`;
    case "heading-level-jump":
      return `标题层级从 H${detail(
        diagnostic,
        "previous_level",
        "?",
      )} 跳到 H${detail(diagnostic, "current_level", "?")}`;
    case "resource-path-escape":
      return `${
        detail(diagnostic, "resource_type") === "image" ? "图片" : "参考文献"
      }路径越出论文资源目录：${diagnostic.target ?? ""}`;
    case "missing-image":
      return `图片不存在：${diagnostic.target ?? ""}`;
    case "missing-bibliography":
      return diagnostic.target
        ? `参考文献文件不存在：${diagnostic.target}`
        : "文档包含引用，但未配置本地 bibliography 路径";
    case "invalid-bibliography":
      return `参考文献数据无效：${diagnostic.target ?? ""}：${detail(
        diagnostic,
        "problem",
        diagnostic.message,
      )}`;
    case "missing-citation":
      return `本地参考文献中不存在 citation key：${diagnostic.target ?? ""}`;
    case "missing-template":
      return `找不到模板：${detail(
        diagnostic,
        "selector",
        diagnostic.target ?? "template",
      )}`;
    case "ambiguous-template":
      return `模板 ID 不唯一：${detail(
        diagnostic,
        "template_id",
        diagnostic.target ?? "",
      )}：${detail(diagnostic, "paths")}`;
    case "invalid-template":
      return `模板无效：${detail(diagnostic, "path")}：${detail(
        diagnostic,
        "field",
        diagnostic.target ?? "",
      )}：${detail(diagnostic, "problem")}`;
    case "missing-template-style":
      return `模板未定义所需样式：${diagnostic.target ?? ""}`;
    default:
      return diagnostic.message;
  }
}

export function presentDiagnostics(
  diagnostics: SerializedDiagnostic[],
): DiagnosticPresentation[] {
  return diagnostics
    .map((diagnostic, index) => ({
      ...diagnostic,
      id: `${diagnostic.code}:${diagnostic.line ?? 0}:${
        diagnostic.target ?? ""
      }:${index}`,
      message: localizedDiagnosticMessage(diagnostic),
    }))
    .sort(
      (left, right) =>
        (left.line ?? -1) - (right.line ?? -1) ||
        severityOrder[left.severity] - severityOrder[right.severity] ||
        compareText(left.code, right.code) ||
        compareText(left.target ?? "", right.target ?? "") ||
        compareText(left.message, right.message),
    );
}

export function diagnosticSummary(
  diagnostics: DiagnosticPresentation[],
): DiagnosticSummary {
  const summary: DiagnosticSummary = {
    all: diagnostics.length,
    error: 0,
    warning: 0,
    info: 0,
  };
  for (const diagnostic of diagnostics) {
    summary[diagnostic.severity] += 1;
  }
  return summary;
}

export function selectVisibleDiagnostics(
  diagnostics: DiagnosticPresentation[],
  filter: DiagnosticFilter,
) {
  return filter === "all"
    ? diagnostics
    : diagnostics.filter((diagnostic) => diagnostic.severity === filter);
}

export function hasFatalDiagnostics(
  diagnostics: DiagnosticPresentation[],
) {
  return diagnostics.some((diagnostic) => diagnostic.severity === "error");
}

File: src/thesis_forge/presentation/diagnostics.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from thesis_forge.core.model import ValidationIssue


def localized_issue_message(issue: ValidationIssue) -> str:
    details = issue.details
    if issue.code == "required-metadata":
        return f"缺少必填元数据：{details.get('path', issue.target or '')}"
    if issue.code == "empty-document":
        return "论文正文为空"
    if issue.code == "invalid-id-prefix":
        return (
            f"ID 前缀无效：{issue.target or ''}，"
            f"期望 {details.get('expected', '')}"
        )
    if issue.code == "duplicate-id":
        return f"重复 ID：{issue.target or ''}"
    if issue.code == "missing-reference":
        return f"引用目标不存在：{issue.target or ''}"
    if issue.code == "heading-level-jump":
        return (
            f"标题层级从 H{details.get('previous_level', '?')} "
            f"跳到 H{details.get('current_level', '?')}"
        )
    if issue.code == "resource-path-escape":
        kind = "图片" if details.get("resource_type") == "image" else "参考文献"
        return f"{kind}路径越出论文资源目录：{issue.target or ''}"
    if issue.code == "missing-image":
        return f"图片不存在：{issue.target or ''}"
    if issue.code == "missing-bibliography":
        if issue.target:
            return f"参考文献文件不存在：{issue.target}"
        return "文档包含引用，但未配置本地 bibliography 路径"
    if issue.code == "invalid-bibliography":
        return (
            f"参考文献数据无效：{issue.target or ''}："
            f"{details.get('problem', issue.message)}"
        )
    if issue.code == "missing-citation":
        return f"本地参考文献中不存在 citation key：{issue.target or ''}"
    if issue.code == "missing-template":
        return f"找不到模板：{details.get('selector', issue.target or 'template')}"
    if issue.code == "ambiguous-template":
        return (
            f"模板 ID 不唯一：{details.get('template_id', issue.target or '')}："
            f"{details.get('paths', '')}"
        )
    if issue.code == "invalid-template":
        return (
            f"模板无效：{details.get('path', '')}："
            f"{details.get('field', issue.target or '')}："
            f"{details.get('problem', '')}"
        )
    if issue.code == "missing-template-style":
        return f"模板未定义所需样式：{issue.target or ''}"
    return issue.message
